### ISSUE TITLE
Add support for build flavours

### ISSUE THREAD

**Issue by snowleopard** [snowleopard/hadrian#188]
The `make`-based build system supports a number of _build flavours_, collections of settings tailored for a particular use-case, e.g. `quick`, `validate`, etc.

Relevant issues: #132, #186, #187.

**Comment by snowleopard** [comment]
> I put build flavours in separate files (ghc/ghc@2c24fd7), so I could more easily `diff` them, and spot inconsistencies. YMMV

@thomie Let me respond to your comment here. 

Yes, for now I am also putting different flavours in separate files. Let's see how small/big these files will eventually become. We probably wouldn't want to multiple files each containing a single useful line and a dozen of boilerplate imports etc., but maybe this won't happen.

**Comment by thomie** [comment]
As discussed in #243, it would be nice to be able to select a BuildFlavour from within `User.hs`.

Other setting in `User.hs` should then override settings predefined by that BuildFlavour. And command-line flags should yet again override all the previous settings.

**Comment by snowleopard** [comment]
The relevant bits are [here](https://github.com/snowleopard/hadrian/blob/master/src/CmdLineFlag.hs#L23-L32):
```haskell
data Flavour      = Default | Quick deriving (Eq, Show)

-- | Default values for 'CmdLineFlag.Untracked'.
defaultUntracked :: Untracked
defaultUntracked = Untracked  { flavour = Default, ... }
```

And [here](https://github.com/snowleopard/hadrian/blob/master/src/Settings/Args.hs#L90-L92):
```haskell
flavourArgs :: Args
flavourArgs = mconcat
    [ cmdFlavour == Quick ? quickFlavourArgs ]

getArgs :: Expr [String]
getArgs = fromDiffExpr $ mconcat [ defaultBuilderArgs
                                 , defaultPackageArgs
                                 , flavourArgs
                                 , userArgs ]
```

The trick is to implement the overriding of defaults in `Settings/User.hs` without introducing import cycles... Not sure how to do that yet.

**Comment by snowleopard** [comment]
I implemented build flavours while working on #268, so I'm closing this issue.
